### PR TITLE
added support to serve a single AudioStream multiple times concurrently through HTTP

### PR DESCRIPTION
**Pull request by kaikreuzer**
Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/FixedLengthAudioStream.java [hunk 1]
```diff
@@ -7,8 +7,10 @@
  */
 package org.eclipse.smarthome.core.audio;
 
+import java.io.InputStream;
+
 /**
- * This is an {@link AudioStream}, which can provide information about its absolute length.
+ * This is an {@link AudioStream}, which can provide information about its absolute length and is able to
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/FixedLengthAudioStream.java]: "... be cloned" or something similar is missing.

File: bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java [hunk 2]
```diff
@@ -37,7 +39,7 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * A servlet that serves audio streams via HTTP.
+ * A servlet that serves audio oneTimeStreams via HTTP.
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: We should keep "streams" here. I assume this has been done by the refactor procedure.

File: bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java [hunk 3]
```diff
@@ -127,32 +140,50 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
-        } else {
-            // TODO: We should also support streams without mark support, but this will mean that we have to read it
-            // first to memory or file system and create a new AudioStream from there.
-            logger.warn("Stream cannot be reset, so it is served without size through HTTP");
-            return serve(stream);
-        }
+    public URL serve(FixedLengthAudioStream stream, int seconds) {
+        String streamId = UUID.randomUUID().toString();
+        multiTimeStreams.put(streamId, stream);
+        streamTimeouts.put(streamId, System.currentTimeMillis() + (seconds * 1000L));
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: I would prefer to use `System.nanoTime()` for such time, as it is independent of the system time and cannot be corrupted by time changes.

File: bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java [hunk 4]
```diff
@@ -127,32 +140,50 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
+            }
+            IOUtils.copy(is, os);
             resp.flushBuffer();
             IOUtils.closeQuietly(stream);
         }
     }
 
+    private synchronized void removeTimedOutStreams() {
+        for (String streamId : multiTimeStreams.keySet()) {
+            if (streamTimeouts.get(streamId) < System.currentTimeMillis()) {
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: See comment below about `System.nanoTime()`

File: bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java [hunk 5]
```diff
@@ -127,32 +140,50 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
+                // the stream has expired, we need to remove it!
+                FixedLengthAudioStream stream = multiTimeStreams.remove(streamId);
+                streamTimeouts.remove(streamId);
+                IOUtils.closeQuietly(stream);
+                stream = null;
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: There is no need to set the local scope reference to null, but if you prefer...

File: bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java [hunk 6]
```diff
@@ -127,32 +140,50 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
+                    is = ((FixedLengthAudioStream) stream).getClonedStream();
+                } catch (AudioException e) {
+                    resp.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.getMessage());
+                }
+            } else {
+                is = stream;
+            }
+            IOUtils.copy(is, os);
             resp.flushBuffer();
             IOUtils.closeQuietly(stream);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: We should close `is` now to ensure the cloned stream if `multiAccess` is true (if it is false `is` is equal to `stream`).

Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/internal/AudioServlet.java]: good catch!

**Comment by kaikreuzer** [comment]
Thanks for the review - PR is updated.

**Pull request merged**